### PR TITLE
provide ability to use custom annotations in page objects

### PR DESCRIPTION
**Pull request by artkoshelev**
There are two major changes in this PR:
- split Annotations into abstract class and default implementation
- extend DefaultElementLocator and AjaxElementLocator constructors to work with provided annotations

### CODE REVIEW

File: java/client/src/org/openqa/selenium/support/pagefactory/DefaultFieldAnnotations.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
+package org.openqa.selenium.support.pagefactory;
```
Comment (andreastt) [inline java/client/src/org/openqa/selenium/support/pagefactory/DefaultFieldAnnotations.java]: Missing copyright header

File: java/client/src/org/openqa/selenium/support/pagefactory/DefaultFieldAnnotations.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import org.openqa.selenium.support.ByIdOrName;
+import org.openqa.selenium.support.CacheLookup;
+import org.openqa.selenium.support.FindAll;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.FindBys;
+
+import java.lang.reflect.Field;
+
+public class DefaultFieldAnnotations extends Annotations {
+	private Field field;
```
Comment (andreastt) [inline java/client/src/org/openqa/selenium/support/pagefactory/DefaultFieldAnnotations.java]: Please reformat this file using the Google Java styleguide.

File: java/client/src/org/openqa/selenium/support/pagefactory/DefaultFieldAnnotations.java (hunk 1 above)
Comment (artkoshelev) [inline java/client/src/org/openqa/selenium/support/pagefactory/DefaultFieldAnnotations.java]: fixed

File: java/client/src/org/openqa/selenium/support/pagefactory/DefaultFieldAnnotations.java (hunk 2 above)
Comment (artkoshelev) [inline java/client/src/org/openqa/selenium/support/pagefactory/DefaultFieldAnnotations.java]: done

**Comment by sevaseva** [comment]
If I remember correctly, this project has a deprecation policy where public APIs are guaranteed to not disappear without having been deprecated for at least a duration of one minor release. This PR deletes some public API and many users' tests will stop compiling.

**Comment by artkoshelev** [comment]
@sevaseva i can rename Annotations -> AbstractAnnotations and DefaultFieldAnnotations -> Annotations, so public API will remain untouched. Is it OK?

**Comment by sevaseva** [comment]
Renaming rename Annotations -> AbstractAnnotations and DefaultFieldAnnotations -> Annotations would help keeping the public APIs untouched, it seems. If you didn't need to change many Selenium test suite tests, that would be a good sign of that. If you like, I could use another big bunch of (proprietary) tests to help you check backward compatibility of a change like that. (if you please could point me to one-commit version of the entire thing on top of a recent-ish head, that would simplify it for me; doesn't have to be a pull request, a commit in your fork would work fine).

Whether or not this whole change is a good idea, I don't have an opinion offhand though. Clearly you are changing an API that wasn't designed with such change in mind; 'AbstractAnnotations' bring a little bit of a smell. I don't have a good feel as to whether the added value this change brings exceeds the harm/smell. I'll defer that decision to others.

**Comment by artkoshelev** [comment]
@sevaseva i rename classes as proposed, squash all recent changes in one commit and rebase it with current master. As you can see, there is no changes in Selenium test suite at all. Test it with additional tests will be great.

**Comment by sevaseva** [comment]
Thanks @artkoshelev. I didn't find any backwards incompatibility. It will still break *some* things; only things that depend on non-public non-APIs, etc; I believe what it will break is okay to break by a useful change.

There are nits you could fix: not fix indentation where it is not broken; use spaces and not tabs, add javadoc to the public classes and methods.

But I have a slightly larger question: why this pull request? I may have missed previous discussion, but I am not sure what becomes possible with this change what isn't possible without it (by subclassing ElementLocatorFactory, for example, it is possible to create a PageFactory supporting custom annotations). Or maybe this change makes certain things just easier?

Such reasoning would be very nice to have in the commit description, for posterity. If it is detailed and verbose - that is not a problem, but a bonus.

**Comment by artkoshelev** [comment]
@sevaseva i've fixed all the nits and added short description to commit.

Implementing recursive location strategy (assume your page object have elements with sub-elements inside) i was forced to duplicate a lot of selenium code. So i realized it's a bed smell and propose this change.

**Comment by sevaseva** [comment]
Thanks @artkoshelev, a few more small comments and I think we can merge this.
Please confirm that you have signed the CLA http://goo.gl/qC50R

**Comment by artkoshelev** [comment]
@sevaseva, I fixed all the bits and made test in both Default- and AjaxElementLocatorTest. Unfortunately providing useful constructive example of custom annotations still means writing a lot of code (mostly because of proxying stuff in DefaultFieldDecorator). I don't think having a whole framework concept in tests is a good idea, but i can add link to [our framework](https://github.com/yandex-qatools/htmlelements) somewhere in javadoc as example how it works. Not sure you need references to any 3-rd party frameworks as well.

**Comment by sevaseva** [comment]
Landed as 4a12de7703942b474ab6b4332750b33a4b899e68 thank you @artkoshelev !

**Pull request closed without merging**